walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.SortedSet;
+import java.util.TreeSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > 1440)
+      return Collections.emptyList();
+
+    // TreeSet has O(log n) insertion time
+    SortedSet<TimeRange> mandatoryBlockers = new TreeSet<>(TimeRange.ORDER_BY_START);
+
+    // which makes this loop an O(n log n) operation
+    events: for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (request.getAttendees().contains(attendee)) {
+          // this is a blocker, split up the available time segment
+          mandatoryBlockers.add(event.getWhen());
+          continue events;
+        }
+      }
+    }
+
+    // mergeTimeRanges is O(n)
+    List<TimeRange> combinedMandatoryBlockers = mergeTimeRanges(mandatoryBlockers);
+
+    // and so is findAvailability
+    return findAvailability(combinedMandatoryBlockers, request.getDuration());
+  }
+
+  private static List<TimeRange> mergeTimeRanges(Collection<TimeRange> ranges) {
+    List<TimeRange> merged = new ArrayList<>(ranges.size());
+    TimeRange prev = null;
+
+    for (TimeRange current : ranges) {
+      // first range should always be added
+      if (prev == null) {
+        prev = current;
+        merged.add(prev);
+        continue;
+      }
+
+      // if the previous range overlaps the current range, extend it
```

Comment:
Comments should declare the intentions of the block, not define what the code does. A developer can read the code and figure out what it does, but can often miss the bigger picture. Here, for example, you may write "timespans overlapping with the previous one need to be merged together". This allows folks to refactor this code in the future without following the exact if statement.